### PR TITLE
ready to merge with master

### PR DESCRIPTION
**Pull request by mmcgoey**
Note 1: I might need to add some changes so that the post is sent to all authors inboxes on the server depending on this response. Currently it only sends a post to an authors inbox if they are a follower or if they are the receiver of the private method. 
https://eclass.srv.ualberta.ca/mod/forum/discuss.php?d=2102414

Note 2: the spec says that you can edit/delete public posts. Since it did not say anything about editing/deleting private and friends posts I did not add that functionality since I would need to update the post in all friends/receivers inboxes

### CODE REVIEW

File: backend/webserver/views.py [hunk 1]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+        return post
+    
+    def get(self,request,pk,post_id, *args, **kwargs):
+        author =self.get_author(pk=pk)
+        post = self.get_post(post_id,author.id)
+        if post.visibility == "PUBLIC":
+            serializer = PostSerializer(post, context={'request': request})
+            return Response(serializer.data, status=status.HTTP_200_OK)
+        else:
+            raise Http404
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]:
```suggestion
            return Response({'message': 'requested post is not public'}, status=status.HTTP_400_BAD_REQUEST)
```
This is an incorrect use of the 404 status code. 404 means that something is not found which incorrect because a post was found, it was just not public. You should return a more meaningful message to the user.

File: backend/webserver/views.py [hunk 2]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+
+    def post(self, request, pk,post_id, *args, **kwargs):
+        author = self.get_author(pk=pk)
+        post = self.get_post(post_id,author.id)
+
+        if post.visibility == "PUBLIC":   
+            if author.id == request.user.id:
+                post = self.get_post(post_id,author.id)
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]: You already fetched the post at line 97. Why are we fetching it again here?

File: backend/webserver/views.py [hunk 3]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+                date_edited = datetime.datetime.utcnow().replace(tzinfo=utc)
+                post.edited_at = date_edited
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]:
```suggestion
                post.edited_at = datetime.datetime.utcnow().replace(tzinfo=utc)
```

File: backend/webserver/views.py [hunk 4]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+                serializer = UpdatePostSerializer(instance=post,data=request.data,partial=True,context={'request': request})
+                if serializer.is_valid():
+                    serializer.save()
```
Comment (zarifmahfuz) [inline backend/webserver/views.py:104]: The `edited_at` column will not get changed because you never saved that. The right way to update the `edited_at` field of a post is by overriding the `serializer.save()` method for `UpdatePostSerializer`. Here is a [sample google search](https://www.google.com/search?q=how+to+override+serializer.save+in+djangorest&rlz=1C5CHFA_enCA966CA966&sxsrf=ALiCzsbQUd9U_j6lBa91av46efLJTEqDJQ%3A1666475400588&ei=iGVUY9bDI8OL0PEP5-OtyAc&ved=0ahUKEwjWp53S6PT6AhXDBTQIHedxC3kQ4dUDCA8&uact=5&oq=how+to+override+serializer.save+in+djangorest&gs_lcp=Cgdnd3Mtd2l6EAMyBwghEKABEAoyCAghEBYQHhAdOgoIABBHENYEELADSgQITRgBSgQIQRgASgQIRhgAUMYGWLUJYN0LaAJwAXgAgAGDAYgBnQOSAQMyLjKYAQCgAQHIAQjAAQE&sclient=gws-wiz).

File: backend/webserver/views.py [hunk 5]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+            if author.id == request.user.id:
+                post = self.get_post(post_id,author.id)
+                post.delete()
+            else:
+                return Response({'message': 'You cannot delete another authors post'}, status=status.HTTP_400_BAD_REQUEST) 
+            return Response({"message":"Object deleted!"}, status=status.HTTP_200_OK)
+        else:
+            return Response({'message': 'You can only delete public posts'}, status=status.HTTP_400_BAD_REQUEST)
+    
+
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]: Why do we have so many spaces after the end of the class?

File: backend/webserver/views.py [hunk 6]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+        post = self.get_post(post_id,author.id)
+        if post.visibility == "PUBLIC":
+            if author.id == request.user.id:
+                post = self.get_post(post_id,author.id)
+                post.delete()
+            else:
+                return Response({'message': 'You cannot delete another authors post'}, status=status.HTTP_400_BAD_REQUEST) 
+            return Response({"message":"Object deleted!"}, status=status.HTTP_200_OK)
+        else:
+            return Response({'message': 'You can only delete public posts'}, status=status.HTTP_400_BAD_REQUEST)
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]:
```suggestion
        return Response({'message': 'You can only delete public posts'}, status=status.HTTP_400_BAD_REQUEST)
```
You don't need the else clause

File: backend/webserver/views.py [hunk 7]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+        else:
+            return Response({'message': 'You can only edit public posts'}, status=status.HTTP_400_BAD_REQUEST)
+        
+
+    def delete(self,request,pk,post_id,*args,**kwargs):
+        author =self.get_author(pk=pk)
+        post = self.get_post(post_id,author.id)
+        if post.visibility == "PUBLIC":
+            if author.id == request.user.id:
+                post = self.get_post(post_id,author.id)
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]: Again, why are we re-fetching the post?

File: backend/webserver/views.py [hunk 8]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+                post.delete()
+            else:
+                return Response({'message': 'You cannot delete another authors post'}, status=status.HTTP_400_BAD_REQUEST) 
+            return Response({"message":"Object deleted!"}, status=status.HTTP_200_OK)
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]: You should move this response line under `post.delete()` for better readability

File: backend/webserver/views.py [hunk 9]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+class AllPosts(APIView):
+    authentication_classes = [BasicAuthentication]
+    permission_classes = [permissions.IsAuthenticated]
+    def get_author(self,pk):
+        author = get_object_or_404(Author,pk=pk)
+        return author
+
+    def get(self, request,pk, *args, **kwargs):
+        author = self.get_author(pk)
+        posts = Post.objects.filter(author=author.id).all()
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]: You can simply do -
```suggestion
        posts = author.posts.all()
```

File: backend/webserver/views.py [hunk 10]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+                    title=serializer.data["title"],
+                    description=serializer.data["description"],
+                    unlisted=serializer.data["unlisted"],
+                    content_type=serializer.data["content_type"],
+                    content=serializer.data["content"],
+                    visibility=serializer.data["visibility"]
+                )
+                if new_post.visibility == "FRIENDS" or new_post.visibility == "PUBLIC":
+                    author_one_followers = author.followed_by_authors
+                    for follow in author_one_followers.iterator():
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]:
```suggestion
                    for follow in author.followed_by_authors.iterator():
```

File: backend/webserver/views.py [hunk 11]
```diff
@@ -69,6 +72,112 @@ def post(self, request):
...
+                    if private_post_serializer.is_valid():
+                        
+                        try:
+                            receiver =  Author.objects.get(pk=private_post_serializer.data['receiver']['id'])    
+                        except Author.DoesNotExist:
+                            return Response({'message': f'receiver author with id {private_post_serializer.data["receiver"]["id"]} does not exist'}, status=status.HTTP_404_NOT_FOUND)
+                        with transaction.atomic():
+                            Inbox.objects.create(target_author=receiver, post=new_post)
+                    else:
+                        print("What are the errors",private_post_serializer.errors)
```
Comment (zarifmahfuz) [inline backend/webserver/views.py]:
```suggestion
```

File: backend/webserver/tests.py [hunk 12]
```diff
@@ -421,3 +424,439 @@ def test_author_does_not_exist(self):
...
+        self.assertEqual(post_1["description"], "Testing post 1")
+        self.assertEqual(post_2["description"], "Testing post 2")
+    
+    def test_send_friend_post_to_inbox(self):
+        author_1 = Author.objects.create(username="author_1", display_name="author_1")
+        author_1.set_password("pass123")
+        author_1.save()
+        request_payload = {"username": "author_1", "password": "pass123"}
+        self.client.post("/login/", data=request_payload, format="json")
+        author_2 = Author.objects.create(username="author_2", display_name="author_2")
```
Comment (zarifmahfuz) [inline backend/webserver/tests.py:891]: You should create more followers here and check if multiple inboxes are updated

File: backend/webserver/tests.py [hunk 13]
```diff
@@ -421,3 +424,439 @@ def test_author_does_not_exist(self):
...
+        response = self.client.post(url,data=payload,format="json")
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        inbox = Inbox.objects.first()
+        self.assertEqual(author_2, inbox.target_author)
+        self.assertEqual(payload["title"],inbox.post.title)
+        self.assertEqual(payload["description"],inbox.post.description)
+        self.assertEqual(payload["unlisted"],inbox.post.unlisted)
+        self.assertEqual(payload["visibility"],inbox.post.visibility)
+        self.assertEqual(payload["content_type"],inbox.post.content_type)
+
```
Comment (zarifmahfuz) [inline backend/webserver/tests.py:950]: Can you think of more test cases? Are you testing all edge cases?

**Review by zarifmahfuz — CHANGES_REQUESTED** [review]
You are close! Remember that you also need to update the API documentation with this PR in the same format that is currently layed out.
